Update ICS 20 readme (#882)
informalsystems/quint: examples/cosmos/ics20/README.md

The ICS 20 protocol itself is made up of a function `sendFungibleTokens` that initiates a token transfer to another chain, and the IBC callback handlers `onRecvPacket`, `onAcknowledgePacket`, and `onTimeoutPacket`. We specify these in [`ics20.qnt`](./ics20.qnt).

Finally, ICS 20 assumes an abstract bank module that is used for minting, burning, and on-chain transfer of fungible tokens. We provide a specification in [`bank.qnt`](./bank.qnt).

## How to use

### Interactively

Quint allows you to interact with the ICS 20 specification, for example to inspect intermediate system states. To do so, start the Quint REPL and load the spec:

```sh
$ cd examples/cosmos/ics20
$ quint repl -r ics20.qnt::ics20Test
Quint REPL 0.10.0
Type ".exit" to exit, or ".help" for more information
true

>>>
```

`ics20Test` sets up three chains `"A"`, `"B"` and `"C"`, all connected to each other via a channel `"channelToX"`.

It also provides an action `init` that sets up the initial state of the protocol. This assigns 100 ATOM to account `"alice"` on chain `"A"`.
The state of each chain is available as `chainStates.get(`chain`)`, and can be inspected interactively:

> ```json

Author: Oh dang, it looked so nice in VSCode, I didn't bother to check on GH... 🤪

```json
>>> chainStates.get("B").bank
Map()
```

Let's now take action `receivePacket`, which triggers receival of the packet:

```json
>>> receivePacket("A", "B")
true
>>> chainStates.get("B")
{
  bank: Map("bob" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 1)),
  channelEscrowAddresses: Map("channelToA" -> "escrow_account", "channelToC" -> "escrow_account"),
  channels: Map("channelToA" -> "channelToB", "channelToC" -> "channelToB"),
  inAcknowledgements: Set(),
  inTimeouts: Set(),
  outPackets: Set(),
  receivedButUnacknowledgedPackets: Set()
}
```

Now Bob has been minted the sent token. We can also see that the token's base denomination has been extended with a trace of its origin (`{ channel: "channelToA", port: "transfer" }`).

Receival of the packet has created an IBC acknowledgement. Although receival of a successful acknowledgement has no effect, let's receive it on chain `"A"`:

```json
>>> chainStates.get("A").inAcknowledgements
Set({ errorMessage: "", success: true })
>>> receiveAck("A")
true
```

> To take the protocol through a complete transfer, in a single step, use the `sendTransfer` action:

Reviewer: `sendTransfer` is actually a run
```suggestion
To take the protocol through a complete transfer, in a single step, use the `sendTransfer` run:
```

It also provides an action `init` that sets up the initial state of the protocol. This assigns 100 ATOM to account `"alice"` on chain `"A"`.
The state of each chain is available as `chainStates.get(`chain`)`, and can be inspected interactively:

```bluespec
>>> init
true
>>> chainStates.get("A")
{
  bank: Map("alice" -> Map({ baseDenom: "atom", path: [] } -> 100)),
  channelEscrowAddresses: Map("channelToB" -> "escrow_account", "channelToC" -> "escrow_account"),
  channels: Map("channelToB" -> "channelToA", "channelToC" -> "channelToA"),
  inAcknowledgements: Set(),
  inTimeouts: Set(),
  outPackets: Set(),
  receivedButUnacknowledgedPackets: Set()
}
```

We can now initiate a transfer of 1 ATOM from `"alice"` on chain `"A"` to `"bob"` on chain `"B"`:

```bluespec
>>> sendPacket("A", "B", ATOM, 1, "alice", "bob")
true
```

Following the ICS 20 protocol, 1 ATOM has been moved out of Alice's account, and into a module-specific escrow account:

```bluespec
>>> chainStates.get("A")
{
  bank: Map(
    "alice" -> Map({ baseDenom: "atom", path: [] } -> 99),
    "escrow_account" -> Map({ baseDenom: "atom", path: [] } -> 1)
  ),

>   [...]

Reviewer 2: ```suggestion
  // ...
```

true
```

To take the protocol through a complete transfer, in a single step, use the `sendTransfer` run:

```bluespec
>>> sendTransfer("A", "B", ATOM, 50, "alice", "charlie")
true
>>> chainStates.get("A").bank
Map(
  "alice" -> Map({ baseDenom: "atom", path: [] } -> 49),
  "escrow_account" -> Map({ baseDenom: "atom", path: [] } -> 51)
)
>>> chainStates.get("B").bank
Map(
  "bob" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 1),
  "charlie" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 50)
)
```

We also provide actions `timeoutPacket` and `receiveTimeout`, that can be called instead of `receivePacket` and `receiveAck` to simulate a timed out transfer.

### Tests

Instead of exploring the protocol interactively, we can encode longer protocol runs, together with assertions that assure correct behavior, as Quint tests. Use the Quint test runner to execute the tests in [`ics20.qnt`](./ics20.qnt):

```sh
$ quint test --main=ics20Test examples/cosmos/ics20/ics20.qnt

  ics20Test
    ok ABCACBATest passed 10000 test(s)
    ok TransferTest passed 10000 test(s)
    ok FailedAckTest passed 10000 test(s)

>     ok TimeoutTest passed 10000 test(s)

Reviewer 2: is there non-determinism inside? I wonder why the engine is running 10000 tests

Author: what constitutes non-determinism?
there's `oneOf` for picking a packet, but in the test the set it is picking out of always has cardinality 1.

  4 passing
```

## Assumptions

Where the [natural language spec][ICS 20] is imprecise or underspecified, we
make the following assumptions:

### `bank.qnt`

> In general, the bank module is underspecified. We provide an abstract,

Reviewer 2: ```suggestion
In general, the bank module is not specified in detail. We provide an abstract,
```

}
```

> ### ICS 20

Reviewer: Is this header meant to have this name? I found it confusing.

Author: Could be confusing, I agree. I was trying to group by file, so I've just replaced it with the literal file name.
